### DISCUSSION TITLE
Where should I start learning CFEngine policy language?

### DISCUSSION THREAD

**Discussion by olehermanse** [cfengine/core#4475] [General]
If you have a new colleague who has never seen CFEngine policy, what is the first resource you point them to, to start learning?

(I know of many, I'm interested to hear which is your favorite).

**Comment by nickanderson** [comment]
This is a list that I commonly provide people with, I update it from time to time. Of course, I also recommend actual in person training (in the before times), there is really no substitute.

# Additional Resources

-   **[Learning CFEngine](https://cf-learn.info/):** Widely considered best book to get a good grounding in CFEngine.
-   **[CFEngine Tutorial](http://www.cfenginetutorial.org/):** Training material published by Aleksey Tsalolikhin from Vertical Sysadmin, one of our partners.
-   **[VSA Training Material](https://www.cfenginetutorial.org):** Self paced in depth tutorial based on Vertical Sysadmin training coursework
-   **[Example Policy Layout](https://github.com/nickanderson/example-a10042):** An example policy layout
-   **[CFEngine Spacemacs Layer](https://github.com/syl20bnr/spacemacs/tree/master/layers/%252Btools/cfengine):** The best editor is neither Emacs nor Vim, it's Emacs and Vim!
-   **[`cf-locate`](https://github.com/cfengine/core/tree/master/contrib/cf-locate):** Find and optionally display body and bundle definitions within a policy set
-   **[Policy Style Guide](https://docs.cfengine.com/docs/3.15/guide-writing-and-serving-policy-policy-style.html):** Write policy in a common style. Make it easier for others to quickly digest your policy.

-   **[`cf-keycrypt`](https://github.com/cfengineers-net/cf-keycrypt):** Tool to encrypt data with CFEngine public keys (in core as of 3.16.0)
    -   Allows to encrypt data for individual hosts using the public key
    -   Can be used with non host keys for "group" encryption
-   **[`cf-profile`](http://www.cfengineers.net/downloads/cfengine-tools-and-utilities/):** Displays execution time summary for cf-agent run to find time consuming polices
-   **[`vim_cf3`](https://github.com/neilhwatson/vim_cf3):** CFEngine 3 vim plugin with Syntax highlighting
-   **[`reindent.pl`](https://github.com/cfengine/core/blob/master/contrib/reindent.pl):** Re-indent CFEngine policy using this script that leverages the excellent cfengine3 mode in Emacs
-   **[Sublime Text 3 CFEngine Beautifier](https://github.com/naksu/cfengine_beautifier):** Automatically reformat CFEngine policy in Sublime Text
-   **[Sublime Text 3 Syntax Highlighter & Snippets](https://github.com/lastops/sublime-cfengine):** CFEngine Syntax highlighting and snippets for Sublime Text

-   **[`cfe-rsplaytime`](https://github.com/lpefferkorn/cfe-rsplaytime) :** Compute splaytime for a given host
-   **[`cfe-profiler`](https://github.com/lpefferkorn/cfe-profiler):** Measures bundle execution time helping to uncover the most time consuming bundles
-   **[Editor support for CFEngine](https://cfengine.com/cfengine-code-editors/):** A list of editors with references for making working with cfengine syntax easier
-   **[CFEngine in a Day](https://www.youtube.com/watch?v=zviHfjKz1nM&t=6s&list=PLh71Vl9YjMajsWxT8zQuRKEPwosG9HdzV&index=2):** Old, but still good 4 part video series ~ 5 hours of [Mark Burgess](https://en.wikipedia.org/wiki/Mark_Burgess_(computer_scientist)) (original author and founder) giving a one day training at [SurfSara](https://www.surf.nl/en/about-surf/subsidiaries/surfsara/)
-   **[CFEngine Tutorial](http://www.cfenginetutorial.org/):** Training material published by Aleksey Tsalolikhin from Vertical Sysadmin, one of our partners.
-   **[General FAQs](https://docs.cfengine.com/docs/3.12/guide-faq.html):** Frequently asked questions
-   **[Hub Admin FAQ](https://docs.cfengine.com/docs/3.12/enterprise-cfengine-guide-hub_administration.html):** Frequently asked questions about CFEngine Enterprise Hub administration
